src/utils/events/eventListeners/eventListenerGenerator.ts

```diff
@@ -32,11 +34,38 @@ export const eventListenerGenerator = async (
     contractAddress ? `filtering Address: ${contractAddress}` : '',
   );
 
-  provider.on(filter, async (log: Log) => {
+  const handleEvent = async (log: Log): Promise<void> => {
     await addEventToQueue(client, clientType, contractAddress, log, provider);
-  });
+  };
+
+  provider.on(filter, handleEvent);
+
+  const listenerRemover = (): void => {
+    verbose(
+      'Removed listener for Event:',
+      eventSignature,
+      contractAddress ? `filtering Address: ${contractAddress}` : '',
+    );
+    provider.off(filter, handleEvent);
+  };
+
+  if (saveRemoverInContext) {
+    const listenerRemoverKey = generateListenerRemoverKey(
+      contractAddress,
+      clientType,
+      eventSignature,
+    );
+    ctx.listenerRemovers[listenerRemoverKey] = listenerRemover;
+  }
 };
 
+/* Generate a unique key for the listenerRemover. */
```

Comment:
When I say "unique", I'm assuming no client type will have two different events with the exact same signature. Therefore this combination of strings will effectively prevent unintended collisions.

Reply:
If we kept a mapping between contract address and event signature, we could check for existing listeners before adding a new one. I experienced these collisions before in dev, because when restarted, the dev environment replays all the events that happened from block 1.

Reply:
When restarting, memory is destroyed and recreated, no? We'd be starting with a fresh context object.